src/ai/backend/agent/server.py

```diff
@@ -1206,179 +1201,224 @@ async def server_main(
-    local_config.container.krunner_volumes = krunner_volumes
-
-    if agent_updates:
-        new_agent_config = local_config.agent.model_copy(update=agent_updates)
-        local_config = local_config.model_copy(update={"agent": new_agent_config})
-
-    etcd_credentials = None
-    if local_config.etcd.user and local_config.etcd.password:
-        etcd_credentials = {
-            "user": local_config.etcd.user,
-            "password": local_config.etcd.password,
+        monitor.prompt = "monitor (agent) >>> "
+        monitor.console_locals["local_config"] = local_config
+        aiomon_started = False
+        try:
+            monitor.start()
+            aiomon_started = True
+        except Exception as e:
+            log.warning(
+                "aiomonitor could not start but skipping this error to continue", exc_info=e
+            )
+        yield monitor
+        if aiomon_started:
+            monitor.close()
+
+    @asynccontextmanager
+    async def etcd_ctx() -> AsyncGenerator[AsyncEtcd]:
+        etcd_credentials = None
+        if local_config.etcd.user and local_config.etcd.password:
+            etcd_credentials = {
+                "user": local_config.etcd.user,
+                "password": local_config.etcd.password,
+            }
+        scope_prefix_map = {
+            ConfigScopes.GLOBAL: "",
+            ConfigScopes.SGROUP: f"sgroup/{local_config.agent.scaling_group}",
+            ConfigScopes.NODE: f"nodes/agents/{local_config.agent.id}",
         }
-    scope_prefix_map = {
-        ConfigScopes.GLOBAL: "",
-        ConfigScopes.SGROUP: f"sgroup/{local_config.agent.scaling_group}",
-        ConfigScopes.NODE: f"nodes/agents/{local_config.agent.id}",
-    }
-    etcd_config_data = local_config.etcd.to_dataclass()
-    etcd = AsyncEtcd(
-        [addr.to_legacy() for addr in etcd_config_data.addrs],
-        local_config.etcd.namespace,
-        scope_prefix_map,
-        credentials=etcd_credentials,
-    )
+        etcd_config_data = local_config.etcd.to_dataclass()
+        etcd = AsyncEtcd(
+            [addr.to_legacy() for addr in etcd_config_data.addrs],
+            local_config.etcd.namespace,
+            scope_prefix_map,
+            credentials=etcd_credentials,
+        )
+        yield etcd
+        await etcd.close()
 
-    rpc_addr = local_config.agent.rpc_listen_addr
-    if not rpc_addr.host:
-        _subnet_hint = await etcd.get("config/network/subnet/agent")
-        subnet_hint = None
-        if _subnet_hint is not None:
-            subnet_hint = ip_network(_subnet_hint)
-        log.debug("auto-detecting agent host")
-        new_rpc_addr = HostPortPair(
-            await identity.get_instance_ip(subnet_hint),
-            rpc_addr.port,
+    @asynccontextmanager
+    async def service_discovery_ctx() -> AsyncGenerator[None]:
+        assert agent_server is not None, (
+            "agent_server must be initialized before the service discovery subsystem."
         )
```

Comment:
I remember we should use `if raise` pattern instead of `assert`.

Reply:
Removed the global variable `agent_server` as the new context-based init routine explicitly expresses the variable dependencies.